### ISSUE TITLE
undocumented retry option: retry_statuses

### ISSUE THREAD

**Issue by Alan-M-Thomaz** [lostisland/faraday#955]
## Basic Info

* Faraday Version: 0.15.4

## Issue description
This option is undocumented.

And i have some questions about a flow using it. I need, in some cases, to retry on status 200 in the same way as i would in status 5xx and timeouts, depending on the response body.
The flow i though is:

> set retry_statuses to [ 200 ]
> set exceptions to [ServerError] + Faraday::Request::Retry::DEFAULT_EXCEPTIONS 
> set methods to [ ], so "retry_request?()" will only depend on my retry_if block (other way it always be true due to the method always be the same in all retries) 
> use retry_if block to check is not a RetryableResponse or check the body if it is 
This should work?

And about this block:
```
raise unless exception.is_a?(Faraday::RetriableResponse)

exception.response
```
that makes it return the normal response, when it is a RetriableResponse, but not retried. Is that what it does?

Also, i'm not sure if is okay thing to do, but theirs any way to change the request body before a retry? (not required for my problem, but would be better, if i could do it)
looks like the body on retries comes is overwrite by "request_body", but it is not passed to retry_block, or other way to edit it, but, i'm not  100% sure about it.

**Comment by Alan-M-Thomaz** [comment]
my code
```
retryable_exceptions = Faraday::Request::Retry::DEFAULT_EXCEPTIONS +[ServerError]
faraday.request  :retry, max: 5, interval: 0.1, interval_randomness: 0.5, backoff_factor: 2,
                   exceptions: retryable_exceptions, retry_statuses: [200],
                   retry_if: Proc.new do |env, exception|
                       return true unless exception.is_a?(Faraday::RetriableResponse)
                       return true if exception.response.body[:results].any? do |result|
                                        result[:error]  == "Unavailable"
                                      end
                    end
```

**Comment by technoweenie** [comment]
> that makes it return the normal response, when it is a RetriableResponse, but not retried. Is that what it does?

Yes, but only after it has attempted to retry in this block:

https://github.com/lostisland/faraday/blob/5063ffe963af083dc76e08a89160e1d71d6c3660/lib/faraday/request/retry.rb#L150-L158

It should retry the request IF:
* Any retries are left (`retries.positive?`)
* If the request method is retriable (any [idempotent methods like GET or HEAD](https://github.com/lostisland/faraday/blob/5063ffe963af083dc76e08a89160e1d71d6c3660/lib/faraday/request/retry.rb#L30)) or if the `retry_if` block returns true (`#retry_request?`)
* If the sleep interval is under the max interval (`#calculate_sleep_amount`)

It looks like it may be retrying every single GET request that returns 200 because it raises `Faraday::RetriableResponse` if the status matches, and `retry_request?` returns true since GET is idempotent. This probably is not what you want. 

## Changing the Request Body

> Also, i'm not sure if is okay thing to do, but theirs any way to change the request body before a retry? 

Not really. There is a `retry_block` option that could enable this:

https://github.com/lostisland/faraday/blob/5063ffe963af083dc76e08a89160e1d71d6c3660/lib/faraday/request/retry.rb#L153

BUT the request body is separated out so it can be reset between retry attempts. This is good if you want to retry a large file upload from the beginning, and not how far the previous request made it before raising an exception.

https://github.com/lostisland/faraday/blob/5063ffe963af083dc76e08a89160e1d71d6c3660/lib/faraday/request/retry.rb#L140-L149

This is bad if you're uploading something like a JSON document that needs something changed between retry attempts (which sounds like your case).

## Suggestion?

The Faraday Retry adapter is designed for simple retry cases in the event of some kind of network or possibly a http protocol issue (if you add 500 to `retry_statuses` or use the `RaiseError` middleware). The problem you describe is more of an application issue that may be too specific for Faraday's generic approach.

You could try implementing this in your app code, but you may not want two retry loops (Faraday Retry middleware's loop looking for exceptions, and your custom loop implementing your `retry_if` block above). Another option is writing a custom middleware combining the JSON (or whatever data format you're using) serializing and the custom retry checks.

**Comment by Alan-M-Thomaz** [comment]
Thx for fast replying, it confirm most things i thought.
 
Just for clarification, cause i may made it a bit confuse, the only real issue was the undocumented about retry_statuses, all the other things were just some questions to be sure i was getting it right, about how this retry works.

>It looks like it may be retrying every single GET request that returns 200 because it raises Faraday::RetriableResponse if the status matches, and retry_request? returns true since GET is idempotent. This probably is not what you want.

You can see the code i commented and deleted? If was that u were talking about, i just didn't emptied the method array, because my case is a POST, wich is already not there.

>BUT the request body is separated out so it can be reset between retry attempts. This is good if you want to retry a large file upload from the beginning, and not how far the previous request made it before raising an exception.

That's what I was afraid of.

>The problem you describe is more of an application issue that may be too specific for Faraday's generic approach.
You could try implementing this in your app code, but you may not want two retry loops.

For now i will just retry the whole request, later, i may try to make my own loop to retry only pieces that are really  needed, when i got more time. It may even become
I'm implementing an retry with exponential backoff for the FCM gem, that recently started to use Faraday.

**Comment by technoweenie** [comment]
> the only real issue was the undocumented about retry_statuses

Thanks for bringing that up. I'll take care of it :)

**Comment by iMacTia** [comment]
Closing as this has been fixed by #957
